### PR TITLE
feat!: `react-query` and supplementary improvements

### LINKED ISSUE #146: feat: using `@tanstack/react-query`
**What is the feature you would like to implement?**
Hooks in `auth-kit` are managing query/mutations states themselves, yet it would be much more reliant to have them cached and managed by `@tanstack/react-query`.

**Why is this feature important?**
Library-independent state management is tough and increases the overall codebase complexity, along with caveat cases that are already managed by `@tanstack/react-query`.

Here's an example:
I have two components, one is a button that creates a channel, and the other one is a UI element that states wether a channel is created or not.

In order to implement one right now, I would have to implement a HOC that would use `useCreateChannel` and I'd need to propagate the state and mutation callback via props to my two components.

If this library leveraged `@tanstack/react-query`, I'd be able to call the hook separately in each component, and as long as the inputs are the same, I'd have the same state from react query cache.

**How should this feature be built? (optional)**
Just need to refactor anything that defines `isSuccess`, `isError`, state finality callback handlers, queries or mutations and such to use `@tanstack/react-uqery`.

**Additional context**
Developers using `auth-kit` would have to define their own `QueryClient`, which they already do with `wagmi@2.0.0`, so likelihood they've done it already is pretty high.

### PR DESCRIPTION
**Pull request by dalechyn**
Will close #146

### CODE REVIEW

**Comment by dalechyn** [comment]
[Date.now()](https://github.com/farcasterxyz/auth-monorepo/blob/b5ad5ec52ba0c6231b939a7c87112fe88ff02173/packages/auth-client/src/clients/transports/http.ts#L87-L89)

This code seems incorrect as the `Date.now()` would be re-evaluated on the `while` clause check. Fixing it as well.

**Comment by dalechyn** [comment]
https://github.com/farcasterxyz/auth-monorepo/blob/b5ad5ec52ba0c6231b939a7c87112fe88ff02173/packages/auth-client/src/actions/app/watchStatus.ts#L19-L31

I believe this is a bit misleading, as we don't *watch* the status change here. In fact, the `poll` function which is used, *polls until success*, meaning that it doesn't return multiple values as an async generator would, but rather retries the request on not satisfying response with a certain interval until a deadline is met.

I'm considering to refactor this as well.

**Comment by dalechyn** [comment]
https://github.com/farcasterxyz/auth-monorepo/blob/b5ad5ec52ba0c6231b939a7c87112fe88ff02173/packages/auth-client/src/messages/verify.ts#L54-L68

This methods should not return any value as they don't mutate it.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -29,16 +29,16 @@ jobs:
         uses: actions/cache@v3
         with:
           path: ${{ github.workspace }}/node_modules
-          key: ${{ runner.os }}--node--${{ hashFiles('**/yarn.lock') }}
+          key: ${{ runner.os }}--node--${{ hashFiles('**/pnpm.lock') }}
```
Comment (dalechyn) [inline .github/workflows/ci.yml:32]: Bad lockfile path

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -60,13 +60,13 @@ jobs:
         uses: actions/cache@v3
         with:
           path: ${{ github.workspace }}/node_modules
-          key: ${{ runner.os }}--node--${{ matrix.node_version}}--${{ hashFiles('**/yarn.lock') }}
+          key: ${{ runner.os }}--node--${{ matrix.node_version}}--${{ hashFiles('**/pnpm.lock') }}
```
Comment (dalechyn) [inline .github/workflows/ci.yml:63]: Bad path

File: .gitignore [hunk 3]
```diff
@@ -1,6 +1,6 @@
 # Logs
-yarn-debug.log*
-yarn-error.log*
+pnpm-debug.log*
```
Comment (dalechyn) [inline .gitignore:2]: Gotta recheck those paths

File: Dockerfile.relay [hunk 4]
```diff
@@ -15,9 +15,9 @@ WORKDIR /home/node/app
 
 # Run turbo prune to create a pruned version of monorepo
 COPY --chown=node:node ./package.json ./package.json
-RUN yarn global add turbo@$(node -e "console.log(require('./package.json').devDependencies.turbo)")
+RUN pnpm global add turbo@$(node -e "console.log(require('./package.json').devDependencies.turbo)")
```
Comment (dalechyn) [inline Dockerfile.relay:18]: Check if this is correct

File: Dockerfile.relay [hunk 5]
```diff
@@ -15,9 +15,9 @@ WORKDIR /home/node/app
...
 COPY --chown=node:node . .
-RUN /home/node/.yarn/bin/turbo prune --scope=@farcaster/auth-relay --docker
+RUN /home/node/.pnpm/bin/turbo prune --scope=@farcaster/auth-relay --docker
```
Comment (dalechyn) [inline Dockerfile.relay:20]: Can use pnpm instead

File: Dockerfile.relay [hunk 6]
```diff
@@ -41,20 +41,20 @@ ENV RUSTFLAGS="-C target-feature=-crt-static"
 
 # Copy dependency information and install all dependencies
 COPY --chown=node:node --from=prune /home/node/app/out/json/ .
-COPY --chown=node:node --from=prune /home/node/app/out/yarn.lock ./yarn.lock
+COPY --chown=node:node --from=prune /home/node/app/out/pnpm.lock ./pnpm.lock
```
Comment (dalechyn) [inline Dockerfile.relay:44]: Bad path

File: packages/auth-kit/src/components/SignInButton/SignInButton.tsx [hunk 7]
```diff
@@ -1,125 +1,121 @@
...
-  const {
+
+  useEffect(() => {
+    if (signInStatus === "success") onSignIn?.(signInData);
+    if (signInStatus === "error") {
+      // if it's a polling error due to the timeout, we recreate the channel
+      if (
+        signInError instanceof AuthClientError &&
+        signInError.errCode === "unavailable" &&
+        signInError.message.startsWith("Polling timed out after")
```
Comment (dalechyn) [inline packages/auth-kit/src/components/SignInButton/SignInButton.tsx:43]: could be a specific `errCode`

File: packages/auth-kit/src/hooks/useSignIn.ts [hunk 8]
```diff
@@ -22,90 +28,84 @@ const defaults = {
...
-    nonce,
-    domain,
-    message: statusData?.message,
-    signature: statusData?.signature,
-    onError,
-  });
+      setSignInMessage({
+        message: pollStatusTillSuccessResponse.message,
+        signature: pollStatusTillSuccessResponse.signature,
+      });
```
Comment (dalechyn) [inline packages/auth-kit/src/hooks/useSignIn.ts:78]: move to useEffect?

File: packages/auth-kit/src/hooks/useSignIn.ts [hunk 9]
```diff
@@ -22,90 +28,84 @@ const defaults = {
...
+        fid: pollStatusTillSuccessResponse.fid,
+        pfpUrl: pollStatusTillSuccessResponse.pfpUrl,
+        username: pollStatusTillSuccessResponse.username,
+        displayName: pollStatusTillSuccessResponse.displayName,
+        bio: pollStatusTillSuccessResponse.bio,
+        custody: pollStatusTillSuccessResponse.custody,
+        verifications: pollStatusTillSuccessResponse.verifications,
+      });
+      return { isAuthenticated, ...pollStatusTillSuccessResponse };
+    },
```
Comment (dalechyn) [inline packages/auth-kit/src/hooks/useSignIn.ts:98]: move to useEffect?

File: examples/with-next-auth/pages/index.tsx [hunk 10]
```diff
@@ -40,38 +40,33 @@ function Content() {
...
-        signature: res.signature,
-        name: res.username,
-        pfp: res.pfpUrl,
-        redirect: false,
-      });
-    },
-    [signIn]
-  );
+  const handleSuccess = useCallback((res: CompletedStatusAPIResponse) => {
+    console.log(res);
```
Comment (horsefacts) [inline examples/with-next-auth/pages/index.tsx:44]:
```suggestion
```

**Comment by dalechyn** [comment]
I'm thinking about the implementation of `useSignInMessage` and `useProfile` hooks, probably they don't need to store the value from the response.

I believe it is a similar state management question as [`wagmi's`](https://github.com/wevm/wagmi/) connector handling.
Specifically, one could say that they could persist the return of the connectors between re-renders, but instead they have the `autoConnect` option that reconnects the providers (although they have cookie persistence to pass previous data until new one is hydrated https://wagmi.sh/react/guides/ssr#persistence-using-cookies).
---
Thus, I believe that `useSignMessage` should not be a client store but rather a function that re-uses query client cache to retrieve the latest used signing message.

For the `useProfile` – I believe it should be a query/a hook that retrieves data from the query client cache as well.

What has to be persisted though is something like a session token, optimistically with a certain deadline.

The current problem is that AFAIK the profile data is retrieved only from the `pollStatusTillSuccess` and is only returned in `CompletedStatusResponse`, therefore is only retrievable when a user undergoes through the QR Code authentication process.

What I believe should be done instead is that `auth-relay` should generate an access token for a user when one sucessfully undergoes the QR Code authentication, and `auth-client` should handle the creation of session tokens, their persistence and expiry.

With those changes applied, hooks for client data retrieval (optimistically `useProfile`) would not require a user to under-go QR flow, but would require an access token that might be persisted by `auth-client` or other client-side management implementations.

**Comment by dalechyn** [comment]
TODO: Refactor use profile and use sign in message hooks.

`useProfile` should make a request to relay server to retrieve the profile as after a page refresh the data is lost.

`useSignInMessage` should take the data from query cache.

**Comment by deodad** [comment]
> I'm thinking about the implementation of `useSignInMessage` and `useProfile` hooks, probably they don't need to store the value from the response.
> 
> I believe it is a similar state management question as [`wagmi's`](https://github.com/wevm/wagmi/) connector handling. Specifically, one could say that they could persist the return of the connectors between re-renders, but instead they have the `autoConnect` option that reconnects the providers (although they have cookie persistence to pass previous data until new one is hydrated https://wagmi.sh/react/guides/ssr#persistence-using-cookies).
> 
> Thus, I believe that `useSignMessage` should not be a client store but rather a function that re-uses query client cache to retrieve the latest used signing message.
> 
> For the `useProfile` – I believe it should be a query/a hook that retrieves data from the query client cache as well.
> 
> What has to be persisted though is something like a session token, optimistically with a certain deadline.
> 
> The current problem is that AFAIK the profile data is retrieved only from the `pollStatusTillSuccess` and is only returned in `CompletedStatusResponse`, therefore is only retrievable when a user undergoes through the QR Code authentication process.
> 
> What I believe should be done instead is that `auth-relay` should generate an access token for a user when one sucessfully undergoes the QR Code authentication, and `auth-client` should handle the creation of session tokens, their persistence and expiry.
> 
> With those changes applied, hooks for client data retrieval (optimistically `useProfile`) would not require a user to under-go QR flow, but would require an access token that might be persisted by `auth-client` or other client-side management implementations.
> 
> cc @horsefacts

I don't think we want the relays to become general purpose auth servers. Session tokens, and their persistence and expiry should be up to the developer and the auth setup they're using.

**Comment by dalechyn** [comment]
> > I'm thinking about the implementation of `useSignInMessage` and `useProfile` hooks, probably they don't need to store the value from the response.
> 
> > 
> 
> > I believe it is a similar state management question as [`wagmi's`](https://github.com/wevm/wagmi/) connector handling. Specifically, one could say that they could persist the return of the connectors between re-renders, but instead they have the `autoConnect` option that reconnects the providers (although they have cookie persistence to pass previous data until new one is hydrated https://wagmi.sh/react/guides/ssr#persistence-using-cookies).
> 
> > 
> 
> > Thus, I believe that `useSignMessage` should not be a client store but rather a function that re-uses query client cache to retrieve the latest used signing message.
> 
> > 
> 
> > For the `useProfile` – I believe it should be a query/a hook that retrieves data from the query client cache as well.
> 
> > 
> 
> > What has to be persisted though is something like a session token, optimistically with a certain deadline.
> 
> > 
> 
> > The current problem is that AFAIK the profile data is retrieved only from the `pollStatusTillSuccess` and is only returned in `CompletedStatusResponse`, therefore is only retrievable when a user undergoes through the QR Code authentication process.
> 
> > 
> 
> > What I believe should be done instead is that `auth-relay` should generate an access token for a user when one sucessfully undergoes the QR Code authentication, and `auth-client` should handle the creation of session tokens, their persistence and expiry.
> 
> > 
> 
> > With those changes applied, hooks for client data retrieval (optimistically `useProfile`) would not require a user to under-go QR flow, but would require an access token that might be persisted by `auth-client` or other client-side management implementations.
> 
> > 
> 
> > cc @horsefacts
> 
> 
> 
> I don't think we want the relays to become general purpose auth servers. Session tokens, and their persistence and expiry should be up to the developer and the auth setup they're using.

Yep! I figured it out a bit later and need to rename some things now.

**Comment by dalechyn** [comment]
I dropped e2e test in the merge conflict on purpose. tbh a bit lazy to rewrite it right now, that effort has to be put later on.

**Comment by dalechyn** [comment]
pipelines fail because I've changed the package manager.

**Comment by dalechyn** [comment]
tbh I have some questions regarding the https://github.com/dalechyn/auth-monorepo/commit/62874a090d0d356964876c752b3ea97975b93e5b but we can think of this later if we want this pr to ever make it lul.

**Comment by dalechyn** [comment]
gonna separate the efforts in this fork: https://github.com/dalechyn/fc-auth, since the changes are too massive to be included.

**Pull request closed without merging**